### PR TITLE
[SYCL][DOC] Add build instructions for Windows to GetStartedWithSYCLC…

### PR DESCRIPTION
**Pull request by v-klochkov**
…ompiler.md

1. Fixed an error in cmake command for Linux.
2. Added steps for OpenCL-Headers and OpenCL-ICD-Loader components, which
are required.
3. Added Windows build instructions.
4. Made the build process more verbose and clear. You can literally follow
the guide step-by-step to build compiler and tools and run a simple test.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

**Comment by v-klochkov** [comment]
See the final result as it appears in browser (not in editor) here: https://github.com/v-klochkov/llvm/blob/public_windows_getting_started/sycl/doc/GetStartedWithSYCLCompiler.md

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 1]
```diff
@@ -16,39 +19,129 @@ to install.
...
+```bash
+cd %WORK_DIR%
+git clone https://github.com/KhronosGroup/OpenCL-Headers
+set OPENCL_HEADERS=%WORK_DIR%\OpenCL-Headers
+```
 
+---
+## Get `OpenCL-ICD-Loader`
+You can also find the most recent instructions for this component at https://github.com/KhronosGroup/OpenCL-ICD-Loader
+### Linux:
```
Comment (mibintc) [inline sycl/doc/GetStartedWithSYCLCompiler.md]: This line should be Linux and Windows:

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 2]
```diff
@@ -16,39 +19,129 @@ to install.
 SYCL support: follow the instructions under
 [SYCL* Compiler and Runtimes](https://github.com/intel/llvm/releases/tag/2019-07)
 
+### Get the required tools:
```
Comment (mibintc) [inline sycl/doc/GetStartedWithSYCLCompiler.md]: the build description refers to ninja so please include that as required tool

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 3]
```diff
@@ -58,12 +151,28 @@ should be used:
...
 Run LIT testing using the command below after building SYCL compiler and runtime.
-
+Use one of these commands:
+### Linux:
 ```bash
 make -j`nproc` check-all
+make -j`nproc` check-llvm
+make -j`nproc` check-llvm-spirv
+make -j`nproc` check-clang
+make -j`nproc` check-sycl
```
Comment (mibintc) [inline sycl/doc/GetStartedWithSYCLCompiler.md]: recently i tried "make all" then "make check-sycl" but check-sycl was not recognized; I got a tip from @alexbatashev that it was necessary to "make sycl-toolchain" before "make check-sycl".  I don't know if this is still needed.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 2 above)
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: 'ninja' is part of MSVC-2017 and MSVC-2019. MSVC is already mentioned in the required tools.
That was one of the reasons why I used 'ninja'.  Using 'make' would require additional downloads/installations.
The 2nd reasons was - ninja is much faster for incremental builds.

Comment (mibintc) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: I didn't know that, nice.  So it looks like cmake is also part of the VS2017 installation?

Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: You would need: 'python.exe' for msvc-2019
'python.exe' + 'cmake.exe' for msvc-2017

Comment (alexbatashev) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: @mibintc @v-klochkov we require CMake 3.2+. I'd rather go with something like `cmake --build . --target sycl-toolchain` and never care about build tool again.

Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: Alex, from your comment I don't understand what you suggest.
Do you suggest only saying explicitly in the doc that cmake must be 'cmake 3.2 or later'?
Or also change the 'cmake' command use below at the section explaining how to build compiler?

Comment (alexbatashev) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: @v-klochkov the latter. I think it would be better to make our instructions generator-agnostic as much as possible. Using `cmake` for build is considered best practice and shows people how to use whatever generator they have/like/etc.

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 4]
```diff
@@ -16,39 +19,129 @@ to install.
...
+
 # Build the SYCL compiler and runtime
 
-Download the LLVM* repository with SYCL support to your local machine folder
-e.g. `$HOME/sycl` (assuming environment var `$SYCL_HOME`) folder using
-following command:
+Download the LLVM repository with SYCL support and required tools to your local machine folder, e.g. $WORK_DIR. 
+For simplicity it is assumed below that environment variable WORK_DIR contains path to an existing folder.
+
+## Set some `environment variables`. For example:
```
Comment (bader) [inline sycl/doc/GetStartedWithSYCLCompiler.md]:
```suggestion
## Configure environment
```

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 5]
```diff
@@ -16,39 +19,129 @@ to install.
...
+### Linux:
+```bash
+export WORK_DIR=/export/home/work_dir
+export SYCL_HOME=$WORK_DIR/sycl_workspace
+```
+### Windows:
+Open a developer command prompt using one of tho methods:
+- Click start menu and search forthe command prompt. So, for MSVC-2017 it is '`x64 Native Tools Command Prompt for VS 2017`'
```
Comment (bader) [inline sycl/doc/GetStartedWithSYCLCompiler.md]:
```suggestion
- Click start menu and search for the command prompt. So, for MSVC-2017 it is '`x64 Native Tools Command Prompt for VS 2017`'
```

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 6]
```diff
@@ -16,39 +19,129 @@ to install.
...
+- run 'cmd' and then '`"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvarsall.bat" x64`'
+```bash
+set WORK_DIR=%USERPROFILE%\work_dir
+set SYCL_HOME=%WORK_DIR%\sycl_workspace
+```
+---
+## Get `OpenCL-Headers`
```
Comment (bader) [inline sycl/doc/GetStartedWithSYCLCompiler.md]: Please, clarify that this is an optional step.
CMake will automatically configure OpenCL headers and ICD loader if OpenCL_INCLUDE_DIR and OpenCL_LIBRARY variables are not set.

BTW, these part of the instructions are not related to Windows enabling, so it would be better to have it committed separately. It should be okay to have a separate commit in this PR as long as it documentation only.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 3 above)
Comment (bader) [inline sycl/doc/GetStartedWithSYCLCompiler.md]: All these new commands are executed as part of `check-all`, so I don't see much value in running both.

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 7]
```diff
@@ -1,12 +1,15 @@
 # Overview
-
 The SYCL* Compiler compiles C++\-based SYCL source files with code for both CPU
 and a wide range of compute accelerators. The compiler uses Khronos*
 OpenCL&trade; API to offload computations to accelerators.
 
+---
+---
+---
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: For better readability. 3 lines separate sections started with top header lines (e.g. "# New topic/section".

3 lines contrast to 1 line separators that I added for readability before smaller sections such as 'Get OpenCL Headers', Get OpenCL-ICD-Loader.

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 8]
```diff
@@ -16,39 +19,129 @@ to install.
 SYCL support: follow the instructions under
 [SYCL* Compiler and Runtimes](https://github.com/intel/llvm/releases/tag/2019-07)
 
+### Get the required tools:
+
+   a. `git` - for downloading the sources
+   
+   b. `cmake` - for building the compiler and tools (Get it at: http://www.cmake.org/download)
+      
+   c. `python` - for building the compiler and running tests (Get it at: https://www.python.org/downloads/release/python-2716/ )
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: python3. simply does not work right now - build process does not finish for clang. Python2 - does the work.
If we fix that issues, the DOC can be updated.

[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: I will add the paths to dst for 'git' and 'MSVC'.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 4 above)
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: Ok, fixed. You'll see the change in the updated version of the commit.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 5 above)
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: Ok, will fix the misprint. thank you.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 3 above)
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: check-all includes all others below, right?
The reason why I added those smaller checks is just to show they exist. New users may not know 'check-sycl' or 'check-llvm-spirv' exists.

It makes sense to clarify that moment somehow in this document.

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 9]
```diff
@@ -16,39 +19,129 @@ to install.
...
-Build the SYCL compiler and runtime following instruction below:
+### Windows:
+```bash
+cd %WORK_DIR%
+git clone https://github.com/KhronosGroup/OpenCL-ICD-Loader
+cd OpenCL-ICD-Loader
+mkdir build
+cd build
+set INCLUDE=%OPENCL_HEADERS%;%INCLUDE%
+cmake -G "Ninja" -DOPENCL_ICD_LOADER_REQUIRE_WDK= ..
```
Comment (alexbatashev) [inline sycl/doc/GetStartedWithSYCLCompiler.md]:
```suggestion
cmake -G "Ninja" -DOPENCL_ICD_LOADER_REQUIRE_WDK=OFF ..
```

Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: Surprisingly it worked as well. Fixed.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 6 above)
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: I did not know that OpenCL-Headers and ICD-Loader parts were optional. Probably because I started SYCL build on some older machine with DLLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=1, etc.

Thank you. I am removing these 2 big blocks now.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 2 above)
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: I still do not understand what you suggest. May need some time for questions and experiments with the new approach.
For now I suggest merging the current version and update the document in a new patch.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: I also removed those triple lines

File: sycl/doc/GetStartedWithSYCLCompiler.md [hunk 10]
```diff
@@ -16,55 +16,102 @@ to install.
...
 ```
+### Windows:
+```bash
+mkdir %SYCL_HOME%
+cd %SYCL_HOME%
+git clone https://github.com/intel/llvm -b sycl
+mkdir %SYCL_HOME%\build
+cd %SYCL_HOME%\build
+
+cmake -G "Ninja" -DCMAKE_BUILD_TYPE=Release -DLLVM_EXTERNAL_PROJECTS="llvm-spirv;sycl" -DLLVM_ENABLE_PROJECTS="clang;llvm-spirv;sycl" -DLLVM_EXTERNAL_SYCL_SOURCE_DIR="%SYCL_HOME%\llvm\sycl" -DLLVM_EXTERNAL_LLVM_SPIRV_SOURCE_DIR="%SYCL_HOME%\llvm\llvm-spirv" -DLLVM_TARGETS_TO_BUILD="X86" -DCMAKE_C_COMPILER=cl -DCMAKE_CXX_COMPILER=cl -DCMAKE_C_FLAGS="/GS" -DCMAKE_CXX_FLAGS="/GS" -DCMAKE_EXE_LINKER_FLAGS="/NXCompat /DynamicBase" -DCMAKE_SHARED_LINKER_FLAGS="/NXCompat /DynamicBase" "%SYCL_HOME%\llvm\llvm"
```
Comment (romanovvlad) [inline sycl/doc/GetStartedWithSYCLCompiler.md]: 1. Is there any reason why we suggest using ninja on windows but make on windows?
2. Why do you set -DLLVM_TARGETS_TO_BUILD="X86" only on windows?

Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: regarding 'ninja'. I answered that above. I had 3 reasons:
a) it is already included into MSVC-2017 and 2019. No need to download/install one more tool - 'make'
b) ninja is much faster for incremental/subsequent builds
c) ninja is recommended at http://clang.llvm.org/get_started.html

I'll check if -DLLVM_TARGETS_TO_BUILD="X86" can be removed from this 'cmake' command.

Comment (romanovvlad) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: 1. I meant why don't we suggest using ninja on Linux as well?

Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: why not ninja on Linux..., well..., a) historically, we use it for so long on Linux and everything works great
b) make on Linux is recommended at http://clang.llvm.org/get_started.html (I don't know if there are any underwater rocks when using ninja on Linux)

Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: Regarding "-DLLVM_TARGETS_TO_BUILD="X86" on Windows:
If you do not set the target, then all targets are built. I suppose, it does not make sense building all of them (it is about +1000 additional sources files to compile).
On Linux where it is not set explicitly now I see:
09:11:22 > llvm-config --targets-built
AArch64 AMDGPU ARM BPF Hexagon Lanai Mips MSP430 NVPTX PowerPC RISCV Sparc SystemZ WebAssembly X86 XCore

Comment (romanovvlad) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: Probably we need update cmake line for linux to include this define.

Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: I agree; will add -DLLVM_TARGETS_TO_BUILD="X86" to Linux cmake comand.

File: sycl/doc/GetStartedWithSYCLCompiler.md (hunk 6 above)
Comment (v-klochkov) [reply sycl/doc/GetStartedWithSYCLCompiler.md]: Now I am not so sure that removing those two optional blocks (download OpenCL_Headers and download/build OpenCL-ICD-Loader) was good idea.
I am going to put them to this guide again.

The reason: even with this patch (that is in code-review now): https://github.com/intel/llvm/pull/528 nmake and make work perfectly without those optional steps/blocks, while 'ninja' does not. Ninja (or our build scripts) have some problems with dependencies. Ninja does not build ICD-Loader before building sycl.lib. SYCL CMakeLists.txt look good to me.

**Comment by v-klochkov** [comment]
I updated the doc (added again those 2 optional steps for Headers and ICD-Loader).

BTW, The doc says:
> clang++ -fsycl simple-sycl-app.cpp -o simple-sycl-app.exe -lOpenCL
Do we really need -lOpenCL now after we've got the PluginInterface. If there are no more direct calls of OpenCL functions, then -lOpenCL is not needed, rihgt?
